### ISSUE TITLE
centroid label button and continuous color-by do not work well together

### ISSUE THREAD

**Issue by bkmartinjr** [chanzuckerberg/cellxgene#1397]
the toolbar centroid label button and continuous color-by don't work well together.

Example failures:

1. if I have a continuous annotation colored by, and click the centroid label button, it does nothing.  Should be disabled if it is actually non-functional.

2. if I have nothing colored by, the label button is enabled, and switches modes, but has no discernible effect when clicked on

3. If I color by a categorical annotation, then click the label button, it works as expected.  Then if I click color-by on a continuous annotation, the labels incorrectly show the previous categorical annotation values.

We need consistent model for labeling when a continuous annotation is colored by, and the button enable/disable state needs to correctly reflect its action.  For example, one  possibility:
* display labels only when a categorical annotation is colored by
* enable the button only when it will have an effect

**Comment by seve** [comment]
Relevant PRs and issues: #1358  #1212

**Comment by seve** [comment]
Could you screencap the third behavior? I believe that's a bug

**Comment by bkmartinjr** [comment]
> Could you screencap the third behavior? I believe that's a bug

Here you go.  It only occurs when the continuous value is a gene

![image](https://user-images.githubusercontent.com/3740062/79381688-35b12a80-7f17-11ea-8e60-8f377acbb7e8.png)

**Comment by liaprins-czi** [comment]
The centroids button should be disabled (and centroids should not show) unless a category is being colored by.

The remaining decision is whether the button should remain in "on" mode beneath the disabled state, even with no centroid labels visible. It might sound like the obvious answer is no it should not be on when no centroids are visible, but there's more to consider:
1) We explicitly state in the tooltip for centroids button "*When a category is colored by*, show labels on the graph" (my emphasis). So this sets expectations for when they'd be visible.
2) If the user turns on annotations it is likely in my mind because they like them as part of their cellxgene environment. We should let them turn it on once (per session as of now) so they don't have to keep turning it on when they go back to coloring-by a category.

It is possible that users use this feature more spottily and infrequently for specific categories that they don't look at that often. Then they'd have to turn off annotations once they colored by a different category anyway, though; in this scenario it would just be after they un-colored-by a category (or more directly, colored-by a gene or continuous) if they didn't turn it off before that.

**TL;DR**
- The centroids button should be disabled (and centroids should not show) unless a category is being colored by.
- If centroids button was on before user stopped coloring by a category (either turned off that category's color-by or switched to coloring by a continuous value), then keep the button in "on" state, "below" the disabled state. Do not show centroids until another category is colored by, at which point centroids button should also be enabled (and on).

**Comment by seve** [comment]
@liaprins-czi thanks for the writeup! I'm going to get set on writing up this fix.  Before I start, do we want to address #1358 in the same breath?

**Comment by liaprins-czi** [comment]
Thanks @seve ! I was wondering about #1358 , too ... I have only heard it from that one user so far, so I'm not sure. The obvious way to fix it for her would be to make the current bug (#3 in Bruce's initial list) a feature, haha. I think there are definite use cases for that, when clusters map to the category labels being labeled by the centroids. In that case the centroids would remain when coloring by anything continuous.But not when nothing is colored by (the disabled yet "on" centroid button state would still happen for both cases). Then if user wants to change to showing a different category's labels, they color-by that category, then by the continuous value again. However, I think it would be better to have a way to indicate which category's labels are on the graph. 

Right now I think if this user's ask is needed by more ppl, we can think about better ways to solve that need, the way I described might not be the best way. How about just don't worry about #1358 yet, and I will try to find out from other users when we talk to them about how / if #1358 would be useful.

**Comment by colinmegill** [comment]
Great point @liaprins-czi. I really like your solution above on this thread for the time being. 

It does sound like there's value in #1358, but the use could accomplish this by subsetting to just certain cells of interest, or selecting only those cells, or toggling back and forth. 

If that isn't good enough and we need #1358, it might mean 3 states on the colorby button:

1. color by this category is off (the usual)
2. color by this category is on (the usual)
3. 'pin' this category as the centroid overlay regardless of whatever else I'm coloring by

**Comment by nh3** [comment]
Sorry to re-open this.

> If I color by a categorical annotation, then click the label button, it works as expected. Then if I click color-by on a continuous annotation, the labels incorrectly show the previous categorical annotation values.

Not sure if other users had complaints against it, for me, I consider this a very handy feature rather than a bug.

>It does sound like there's value in #1358, but the user could accomplish this by subsetting to just certain cells of interest, or selecting only those cells, or toggling back and forth.

In practice, it takes much more effort and time clicking back and forth and still doesn't achieve as good a effect as overlaying category labels on top of colouring by continuous variables.
